File: golang/divide_conquer/convert_sorted.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func sortedArrayToBST(nums []int) *TreeNode {
	if len(nums) == 0 {
		return nil
	}

	mid := len(nums) / 2

	root := &TreeNode{
		Val: nums[mid],
	}

	root.Left = sortedArrayToBST(nums[:mid])
	root.Right = sortedArrayToBST(nums[mid+1:])

	return root
}

func preorderTraversal(root *TreeNode) {
	if root == nil {
		return
	}

	fmt.Print(root.Val, " ")
	preorderTraversal(root.Left)
	preorderTraversal(root.Right)
}

File: python/divide_conquer/convert_sorted.py
from typing import List


class TreeNode:
    def __init__(self, val: int = 0, left: 'TreeNode' = None, right: 'TreeNode' = None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> TreeNode | None:
        if len(nums) == 0:
            return None

        mid = len(nums) // 2

        root = TreeNode(
            val=nums[mid]
        )

        root.left = self.sortedArrayToBST(nums=nums[:mid])
        root.right = self.sortedArrayToBST(nums[mid+1:])

        return root


    def preorderTraversal(self,root: TreeNode):
        if root == None:
            return

        print(root.val, " ")

        self.preorderTraversal(root.left)
        self.preorderTraversal(root.right)
